Issue #6 fixed and reading functions added

zaidMughal opened the pull request:
Issue #6 was that createchar() function leave data address in CGRAM position. After that, begin(), setCursor(),clear() or home() functions must be executed before write() function in order for it to work. It is fixed but require rw pin to be set for it to work. This is because, when address register goes to CGRAM, there is no record of its previous position in our MCU. MCU must read its previous position in order for it to go back. In case rw pin is not specified, the register is not returned and it cannot untill rw is specified.

Also Added following functions,
read_BF_addr
read_RAM
read
read4bits
read8bits

zaidMughal committed:
Added reading functions

zaidMughal committed:
Added reading functions and fixed issue #6

Issue # 6 was that createchar() function leave data address in CGRAM position. After that, begin(), setCursor(),clear() or home() functions must be executed before write() function in order for it to work. It is fixed but require rw pin to be set for it to work. This is because, when address register goes to CGRAM, there is no record of its previous position in our MCU. MCU must read its previous position in order for it to go back. In case rw pin is not specified, the register is not returned and it cannot untill rw is specified.
Also Added following functions,
read_BF_addr
read_RAM
read
read4bits
read8bits

zaidMughal committed:
Added reading functions

zaidMughal commented:
Issue # 6 was that createchar() function leave data address in CGRAM position. After that, begin(), setCursor(),clear() or home() functions must be executed before write() function in order for it to work. It is fixed but require rw pin to be set for it to work. This is because, when address register goes to CGRAM, there is no record of its previous position in our MCU. MCU must read its previous position in order for it to go back. In case rw pin is not specified, the register is not returned and it cannot untill rw is specified.
Also Added following functions,
read_BF_addr
read_RAM
read
read4bits
read8bits

zaidMughal commented:
Issue # 6 was that createchar() function leave data address in CGRAM position. After that, begin(), setCursor(),clear() or home() functions must be executed before write() function in order for it to work. It is fixed but require rw pin to be set for it to work. This is because, when address register goes to CGRAM, there is no record of its previous position in our MCU. MCU must read its previous position in order for it to go back. In case rw pin is not specified, the register is not returned and it cannot untill rw is specified.
Also Added following functions,
read_BF_addr
read_RAM
read
read4bits
read8bits

zaidMughal commented:
Issue # 6 was that createchar() function leave data address in CGRAM position. After that, begin(), setCursor(),clear() or home() functions must be executed before write() function in order for it to work. It is fixed but require rw pin to be set for it to work. This is because, when address register goes to CGRAM, there is no record of its previous position in our MCU. MCU must read its previous position in order for it to go back. In case rw pin is not specified, the register is not returned and it cannot untill rw is specified.
Also Added following functions,
read_BF_addr
read_RAM
read
read4bits
read8bits

per1234 commented on src/LiquidCrystal.h:
```diff
@@ -82,12 +82,17 @@ class LiquidCrystal : public Print {
   virtual size_t write(uint8_t);
   void command(uint8_t);
+  uint8_t read_BF_addr(void);
+  uint8_t read_RAM(void);
```
The other functions in the library use camelCase names. Is there a reason you departed from that naming convention for these functions?

Please add the function names added to the public API to [keywords.txt](https://github.com/arduino-libraries/LiquidCrystal/blob/master/keywords.txt), using a true tab as the field separator:
https://arduino.github.io/arduino-cli/library-specification/#keywordstxt-format

zaidMughal replied:
Thanks for pointing out my mistakes. I have corrected the case and added function names to keywords.
Also I note that command function keyword was missing. Was there a reason or should I add it.

per1234 replied:
>I note that command function keyword was missing. Was there a reason

Probably they just forgot to add it. As you probably noticed, it's an easy thing to miss since keywords are purely cosmetic.

>should I add it

I think it would be a good idea (but I'm not the person who decides which pull requests are merged). Please submit that as a separate pull request though, since it's unrelated to the change you propose here.

per1234 replied:
I replied here earlier, but I think I accidentally deleted it, so I'll try again:

>I note that command function keyword was missing. Was there a reason

Probably they just forgot to add it. As you might have noticed, it's easy to do that since keywords are purely cosmetic.

>should I add it

I think it would be nice (though I'm not the one who decides which pull requests get merged). If you want to add it, please do so in a separate pull request, since that change is unrelated to the one you propose in this pull request.

per1234 commented on src/LiquidCrystal.cpp:
```diff
@@ -300,6 +310,35 @@ void LiquidCrystal::send(uint8_t value, uint8_t mode) {
+      address = read4bits();
+    }
+
+     for (int i=0; i<((_displayfunction & LCD_8BITMODE) ? 8 : 4); ++i)
```
```suggestion
    for (int i=0; i<((_displayfunction & LCD_8BITMODE) ? 8 : 4); ++i)
```
Fix indentation

zaidMughal replied:
Thanks again!
Indentation is fixed now.

zaidMughal committed:
camelCase corrected

camelCase corrected in two functions.

zaidMughal committed:
camelCase and indentation correction

zaidMughal committed:
read functions keywords added

zaidMughal committed:
Typo correction

zaidMughal committed:
read functions returned type changed to int

zaidMughal committed:
read function return type changed to int

read function return type changed to int so they give -1 if not read.
Also readRAM generates an error if both inlined and used in cpp file. Either we put it in .h file or not make it a normal function. So I  made it a normal function.

per1234 commented on keywords.txt:
```diff
@@ -31,6 +31,8 @@ scrollDisplayLeft	KEYWORD2
 createChar	KEYWORD2
 setRowOffsets	KEYWORD2
+readBFAddr  KEYWORD2
+readRAM KEYWORD2
```
```suggestion
readBFAddr	KEYWORD2
readRAM	KEYWORD2
```
The Arduino IDE will only recognize the keyword if you use a true tab as a field separator, not spaces.

zaidMughal replied:
I did use tabs but my Github's editor settings were changed.
I have corrected them.

zaidMughal committed:
spaces changed to tabs

per1234 commented on keywords.txt:
```diff
@@ -11,7 +11,7 @@ LiquidCrystal	KEYWORD1	LiquidCrystal
 # Methods and Functions (KEYWORD2)
 #######################################
-
+	
```
```suggestion

```

zaidMughal replied:
done

zaidMughal committed:
extra tab deleted

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino-libraries/LiquidCrystal?pullRequest=32) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino-libraries/LiquidCrystal?pullRequest=32) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino-libraries/LiquidCrystal?pullRequest=32) it.</sub>

zaidMughal commented:
@per1234 
I have signed the CLA.
